### PR TITLE
Sdit 993 map nomis tables and create nomis api endpoints

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderNonAssociation.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      JoinColumnOrFormula(
+        formula = JoinFormula(
+          value = "'" + NonAssociationReason.DOMAIN + "'",
+          referencedColumnName = "domain",
+        ),
+      ),
+      JoinColumnOrFormula(column = JoinColumn(name = "NS_REASON_CODE", referencedColumnName = "code")),
+    ],
+  )
+  var nonAssociationReason: NonAssociationReason? = null,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderNonAssociation.kt:57]: Is this really ever null in real life?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/CreateNonAssociationRequest.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  @Schema(description = "Reason code of the first prisoner, domain NON_ASSO_RSN", required = true, example = "VIC")
+  val reason: String,
+
+  @Schema(description = "Reason code of the other prisoner, domain NON_ASSO_RSN", required = true, example = "PER")
+  val recipReason: String,
+
+  @Schema(description = "Type code, domain NON_ASSO_TYP", required = true, example = "WING")
+  val type: String,
+
+  val authorisedBy: String? = null,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/CreateNonAssociationRequest.kt:27]: No schema?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/NonAssociationResource.kt [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Forbidden, requires role NOMIS_NON_ASSOCIATIONS",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+    ],
+  )
+  fun cancelNonAssociation(
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/NonAssociationResource.kt:152]: Better called `closeNonAssociation`?

**Review by mikehalmamoj — APPROVED** [review]

**Comment by mikehalmamoj** [comment]
That's a lot of untested code - I'm not sure how we're going to know if you've missed any endpoints in the follow up PRs. Maybe do some test coverage to check? It's a lot easier to review with 1 endpoints and its tests in a single PR...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderNonAssociation.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      JoinColumnOrFormula(
+        formula = JoinFormula(
+          value = "'" + NonAssociationReason.DOMAIN + "'",
+          referencedColumnName = "domain",
+        ),
+      ),
+      JoinColumnOrFormula(column = JoinColumn(name = "RECIP_NS_REASON_CODE", referencedColumnName = "code")),
+    ],
+  )
+  var recipNonAssociationReason: NonAssociationReason? = null,
```
[the remark this answers is not in the document]
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderNonAssociation.kt:71]: yep, also nullable in the DB.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/NonAssociationService.kt [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+    OffenderNonAssociationDetail(
+      id = OffenderNonAssociationDetailId(
+        offender = existing.id.offender,
+        nsOffender = existing.id.nsOffender,
+        typeSequence = typeSequence,
+      ),
+      offenderBooking = existing.offenderBooking,
+      nsOffenderBooking = existing.nsOffenderBooking,
+      nonAssociationReason = reason,
+      // recipNonAssociationReason = existing.recipNonAssociationReason,
```
[the remark this answers is not in the document]
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/nonassociations/NonAssociationService.kt:204]: removed

**Pull request merged**